File: ARE.Core/Rules/Rule.cs
using ARE.Core.Abstractions;
using ARE.Core.Abstractions.Enums;
using ARE.Core.Conditions;
using ARE.Core.Core;
using System;
using System.Collections.Generic;

namespace ARE.Core.Rules
{
    public class Rule : IRule
    {
        public string RuleId { get; private set; } = "";

        public string? Group { get; private set; }

        public int Priority { get; private set; }

        public bool IsEnabled { get; private set; } = true;

        public IReadOnlyList<string> EventTypes => _eventTypes;

        public IReadOnlyList<ICondition> Conditions => _conditions;

        public MatchMode MatchMode { get; private set; } = MatchMode.All;

        public IReadOnlyList<ActionBinding> Actions => _actions;

        private readonly List<string> _eventTypes = new();

        private readonly List<ICondition> _conditions = new();

        private readonly List<ActionBinding> _actions = new();

        private Rule() { }

        // ---- Fluent Builder ----
        public static Rule Create(string ruleId) => new() { RuleId = ruleId };

        public Rule InGroup(string group) { Group = group; return this; }

        public Rule WithPriority(int priority) { Priority = priority; return this; }

        public Rule Enable(bool enabled = true) { IsEnabled = enabled; return this; }

        public Rule Disable() { IsEnabled = false; return this; }

        /// <summary>Bu kural hangi event tipini dinlesin</summary>
        public Rule On(string eventType)
        {
            if (!_eventTypes.Contains(eventType))
                _eventTypes.Add(eventType);

            return this;
        }

        /// <summary>Birden fazla event tipi dinle</summary>
        public Rule On(params string[] eventTypes)
        {
            foreach (var et in eventTypes) On(et);

            return this;
        }

        /// <summary>Koşul eşleşme modunu belirle</summary>
        public Rule WithMatchMode(MatchMode mode) { MatchMode = mode; return this; }

        /// <summary>Lambda koşul ekle</summary>
        public Rule When(string name, Func<IEvent, AreContext, bool> predicate)
        {
            _conditions.Add(new DataCondition(name, predicate));

            return this;
        }

        /// <summary>Alan karşılaştırma koşulu ekle</summary>
        public Rule WhenField(string field, CompareOp op, object expected)
        {
            _conditions.Add(new FieldCondition(field, op, expected));

            return this;
        }

        /// <summary>Kısa yol: alan == değer</summary>
        public Rule WhenEquals(string field, object value) => WhenField(field, CompareOp.Equal, value);

        /// <summary>Kısa yol: alan > değer</summary>
        public Rule WhenGreaterThan(string field, object value) => WhenField(field, CompareOp.GreaterThan, value);

        /// <summary>Kısa yol: alan &lt; değer</summary>
        public Rule WhenLessThan(string field, object value) => WhenField(field, CompareOp.LessThan, value);

        /// <summary>Özel ICondition ekle</summary>
        public Rule When(ICondition condition)
        {
            _conditions.Add(condition);

            return this;
        }

        /// <summary>Action bağla (ayarsız)</summary>
        public Rule Then(string actionType, int order = 0)
        {
            _actions.Add(new ActionBinding
            {
                ActionType = actionType,
                Order = order
            });

            return this;
        }

        /// <summary>Action bağla (ayarlı)</summary>
        public Rule Then(string actionType, Action<ActionSettings> configure, int order = 0)
        {
            var settings = new ActionSettings();

            configure(settings);

            _actions.Add(new ActionBinding
            {
                ActionType = actionType,
                Settings = settings,
                Order = order
            });

            return this;
        }
    }
}

File: are_engine_core/lib/src/rule.dart
import 'interfaces.dart';
import 'are_context.dart';
import 'action_settings.dart';
import 'enums.dart';
import 'conditions.dart';

/// Fluent builder for defining rules.
class Rule implements IRule {
  @override
  final String ruleId;

  @override
  String? group;

  @override
  int priority = 0;

  @override
  bool isEnabled = true;

  @override
  final List<String> eventTypes = [];

  @override
  final List<ICondition> conditions = [];

  @override
  MatchMode matchMode = MatchMode.all;

  @override
  final List<ActionBinding> actions = [];

  Rule._(this.ruleId);

  /// Create a new rule with the given ID.
  static Rule create(String ruleId) => Rule._(ruleId);

  /// Assign this rule to a group.
  Rule inGroup(String group) {
    this.group = group;
    return this;
  }

  /// Set the priority (higher = runs first).
  Rule withPriority(int priority) {
    this.priority = priority;
    return this;
  }

  /// Enable this rule.
  Rule enable() {
    isEnabled = true;
    return this;
  }

  /// Disable this rule.
  Rule disable() {
    isEnabled = false;
    return this;
  }

  /// Listen for one or more event types.
  Rule on(String eventType, [String? eventType2, String? eventType3]) {
    if (!eventTypes.contains(eventType)) eventTypes.add(eventType);
    if (eventType2 != null && !eventTypes.contains(eventType2)) {
      eventTypes.add(eventType2);
    }
    if (eventType3 != null && !eventTypes.contains(eventType3)) {
      eventTypes.add(eventType3);
    }
    return this;
  }

  /// Set the condition matching mode.
  Rule withMatchMode(MatchMode mode) {
    matchMode = mode;
    return this;
  }

  /// Add a lambda condition.
  Rule when(String name, bool Function(IEvent evt, AreContext ctx) predicate) {
    conditions.add(DataCondition(name, predicate));
    return this;
  }

  /// Add a custom ICondition.
  Rule whenCondition(ICondition condition) {
    conditions.add(condition);
    return this;
  }

  /// Add a field comparison condition.
  Rule whenField(String field, CompareOp op, dynamic expected) {
    conditions.add(FieldCondition(field, op, expected));
    return this;
  }

  /// Shortcut: field == value.
  Rule whenEquals(String field, dynamic value) =>
      whenField(field, CompareOp.equal, value);

  /// Shortcut: field > value.
  Rule whenGreaterThan(String field, dynamic value) =>
      whenField(field, CompareOp.greaterThan, value);

  /// Shortcut: field < value.
  Rule whenLessThan(String field, dynamic value) =>
      whenField(field, CompareOp.lessThan, value);

  /// Bind an action (without settings).
  Rule then(String actionType, [void Function(ActionSettings s)? configure, int order = 0]) {
    final settings = ActionSettings();
    if (configure != null) configure(settings);
    actions.add(ActionBinding(
      actionType: actionType,
      settings: settings,
      order: order,
    ));
    return this;
  }
}
